File: java/src/test/java/com/google/cloud/crypto/tink/RegistryTest.java
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.cloud.crypto.tink;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import com.google.cloud.crypto.tink.AesCtrHmacAeadProto.AesCtrHmacAeadKey;
import com.google.cloud.crypto.tink.AesCtrHmacAeadProto.AesCtrHmacAeadKeyFormat;
import com.google.cloud.crypto.tink.AesGcmProto.AesGcmKey;
import com.google.cloud.crypto.tink.AesGcmProto.AesGcmKeyFormat;
import com.google.cloud.crypto.tink.CommonProto.HashType;
import com.google.cloud.crypto.tink.HmacProto.HmacKey;
import com.google.cloud.crypto.tink.HmacProto.HmacKeyFormat;
import com.google.cloud.crypto.tink.TestUtil.DummyAead;
import com.google.cloud.crypto.tink.TinkProto.KeyData;
import com.google.cloud.crypto.tink.TinkProto.KeyStatusType;
import com.google.cloud.crypto.tink.TinkProto.KeyTemplate;
import com.google.cloud.crypto.tink.TinkProto.Keyset;
import com.google.cloud.crypto.tink.TinkProto.OutputPrefixType;
import com.google.cloud.crypto.tink.aead.AeadFactory;
import com.google.cloud.crypto.tink.mac.MacFactory;
import com.google.cloud.crypto.tink.subtle.AesGcmJce;
import com.google.cloud.crypto.tink.subtle.EncryptThenAuthenticate;
import com.google.cloud.crypto.tink.subtle.MacJce;
import com.google.protobuf.ByteString;
import com.google.protobuf.MessageLite;
import java.security.GeneralSecurityException;
import java.util.List;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/**
 * Tests for Registry.
 */
@RunWith(JUnit4.class)
public class RegistryTest {
  private static class CustomAeadKeyManager implements KeyManager<Aead, MessageLite, MessageLite> {
    public CustomAeadKeyManager() {}
    @Override
    public Aead getPrimitive(ByteString proto) throws GeneralSecurityException {
      return new DummyAead();
    }
    @Override
    public Aead getPrimitive(MessageLite proto) throws GeneralSecurityException {
      return new DummyAead();
    }
    @Override
    public MessageLite newKey(ByteString template) throws GeneralSecurityException {
      throw new GeneralSecurityException("Not Implemented");
    }
    @Override
    public MessageLite newKey(MessageLite template) throws GeneralSecurityException {
      throw new GeneralSecurityException("Not Implemented");
    }
    @Override
    public KeyData newKeyData(ByteString serialized) throws GeneralSecurityException {
      throw new GeneralSecurityException("Not Implemented");
    }
    @Override
    public boolean doesSupport(String typeUrl) {  // supports same keys as AesGcmKey
      return typeUrl.equals("type.googleapis.com/google.cloud.crypto.tink.AesGcmKey");
    }
  }

  private String aesCtrHmacAeadTypeUrl =
      "type.googleapis.com/google.cloud.crypto.tink.AesCtrHmacAeadKey";
  private String aesGcmTypeUrl =
      "type.googleapis.com/google.cloud.crypto.tink.AesGcmKey";
  private String hmacKeyTypeUrl =
      "type.googleapis.com/google.cloud.crypto.tink.HmacKey";

  @Before
  public void setUp() throws GeneralSecurityException {
    AeadFactory.registerStandardKeyTypes();
    MacFactory.registerStandardKeyTypes();
  }

  @Test
  public void testKeyManagerRegistration() throws Exception {
    // Retrieve some key managers.
    KeyManager<Aead, AesCtrHmacAeadKey, AesCtrHmacAeadKeyFormat> aesCtrHmacAeadManager =
        Registry.INSTANCE.getKeyManager(aesCtrHmacAeadTypeUrl);
    assertTrue(aesCtrHmacAeadManager.getClass().toString().contains(
        "AesCtrHmacAeadKeyManager"));

    KeyManager<Aead, AesGcmKey, AesGcmKeyFormat> aesGcmManager =
        Registry.INSTANCE.getKeyManager(aesGcmTypeUrl);
    assertTrue(aesGcmManager.getClass().toString().contains(
        "AesGcmKeyManager"));

    KeyManager<Mac, HmacKey, HmacKeyFormat> hmacManager =
        Registry.INSTANCE.getKeyManager(hmacKeyTypeUrl);
    assertTrue(hmacManager.getClass().toString().contains(
        "HmacKeyManager"));

    // TODO(thaidn): make this assignment throw some exception.
    KeyManager<Aead, HmacKey, HmacKeyFormat> wrongType =
        Registry.INSTANCE.getKeyManager(hmacKeyTypeUrl);
    assertTrue(wrongType.getClass().toString().contains(
        "HmacKeyManager"));
    KeyTemplate template = TestUtil.createHmacKeyTemplate(
        16 /* key size */, 16 /* tag size */, HashType.SHA256);
    HmacKey hmacKey = Registry.INSTANCE.newKey(template);
    try {
      Aead unused = wrongType.getPrimitive(hmacKey);
      fail("Expected ClassCastException");
    } catch (ClassCastException e) {
      assertTrue(e.toString().contains(
          "MacJce cannot be cast to com.google.cloud.crypto.tink.Aead"));
    }

    String badTypeUrl = "bad type URL";
    try {
      KeyManager<Mac, MessageLite, MessageLite> unused =
          Registry.INSTANCE.getKeyManager(badTypeUrl);
      fail("Expected GeneralSecurityException.");
    } catch (GeneralSecurityException e) {
      assertTrue(e.toString().contains("unsupported"));
      assertTrue(e.toString().contains(badTypeUrl));
    }
  }

  @Test
  public void testKeyAndPrimitiveCreation() throws Exception {
    // Create some keys and primitives.
    KeyTemplate template =  TestUtil.createAesGcmKeyTemplate(16 /* key_size */);
    AesGcmKey aesGcmKey = Registry.INSTANCE.newKey(template);
    assertEquals(16, aesGcmKey.getKeyValue().size());
    KeyData aesGcmKeyData = Registry.INSTANCE.newKeyData(template);
    assertEquals(aesGcmTypeUrl, aesGcmKeyData.getTypeUrl());
    Aead aead = Registry.INSTANCE.getPrimitive(aesGcmKeyData);
    // This might break when we add native implementations.
    assertEquals(AesGcmJce.class, aead.getClass());

    template = TestUtil.createHmacKeyTemplate(
        16 /* key size */, 10 /* tag size */, HashType.SHA256);
    HmacKey hmacKey = Registry.INSTANCE.newKey(template);
    assertEquals(16, hmacKey.getKeyValue().size());
    assertEquals(10, hmacKey.getParams().getTagSize());
    assertEquals(HashType.SHA256, hmacKey.getParams().getHash());
    KeyData hmacKeyData = Registry.INSTANCE.newKeyData(template);
    assertEquals(hmacKeyTypeUrl, hmacKeyData.getTypeUrl());
    Mac mac = Registry.INSTANCE.getPrimitive(hmacKeyData);
    // This might break when we add native implementations.
    assertEquals(MacJce.class, mac.getClass());

    // Create a keyset, and get a PrimitiveSet.
    KeyTemplate template1 =  TestUtil.createAesGcmKeyTemplate(16 /* key_size */);
    KeyTemplate template2 =  TestUtil.createAesCtrHmacAeadKeyTemplate(
        16 /* AES key size */,
        12 /* IV size */,
        16 /* HMAC key size */,
        10 /* tag size */);
    KeyData key1 = Registry.INSTANCE.newKeyData(template1);
    KeyData key2 = Registry.INSTANCE.newKeyData(template1);
    KeyData key3 = Registry.INSTANCE.newKeyData(template2);
    KeysetHandle keysetHandle = new KeysetHandle(Keyset.newBuilder()
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key1)
            .setKeyId(1)
            .setStatus(KeyStatusType.ENABLED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key2)
            .setKeyId(2)
            .setStatus(KeyStatusType.ENABLED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key3)
            .setKeyId(3)
            .setStatus(KeyStatusType.ENABLED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .setPrimaryKeyId(2)
        .build());
    PrimitiveSet<Aead> aeadSet = Registry.INSTANCE.getPrimitives(keysetHandle);
    assertEquals(AesGcmJce.class, aeadSet.getPrimary().getPrimitive().getClass());

    // Try a keyset with some keys non-ENABLED.
    keysetHandle = new KeysetHandle(Keyset.newBuilder()
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key1)
            .setKeyId(1)
            .setStatus(KeyStatusType.DESTROYED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key2)
            .setKeyId(2)
            .setStatus(KeyStatusType.DISABLED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key3)
            .setKeyId(3)
            .setStatus(KeyStatusType.ENABLED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .setPrimaryKeyId(3)
        .build());
    aeadSet = Registry.INSTANCE.getPrimitives(keysetHandle);
    assertEquals(EncryptThenAuthenticate.class,
        aeadSet.getPrimary().getPrimitive().getClass());
  }


  @Test
  public void testRegistryCollisions() throws Exception {
    try {
      Registry.INSTANCE.registerKeyManager(aesCtrHmacAeadTypeUrl, null);
      fail("Expected NullPointerException.");
    } catch (NullPointerException e) {
      assertTrue(e.toString().contains("must be non-null"));
    }
    // This should not overwrite the existing manager.
    assertFalse(Registry.INSTANCE.registerKeyManager(aesCtrHmacAeadTypeUrl,
        new CustomAeadKeyManager()));
    KeyManager<Aead, MessageLite, MessageLite> manager =
       Registry.INSTANCE.getKeyManager(aesCtrHmacAeadTypeUrl);
    assertNotEquals(CustomAeadKeyManager.class, manager.getClass());
    assertTrue(manager.getClass().toString().contains(
        "AesCtrHmacAeadKeyManager"));
  }

  @Test
  public void testInvalidKeyset() throws Exception {
    // Empty keyset.
    try {
      Registry.INSTANCE.getPrimitives(new KeysetHandle(Keyset.newBuilder().build()));
      fail("Invalid keyset. Expect GeneralSecurityException");
    } catch (GeneralSecurityException e) {
      assertTrue(e.toString().contains("empty keyset"));
    }

    // Create a keyset.
    KeyTemplate template1 =  TestUtil.createHmacKeyTemplate(
        16 /* key size */, 10 /* tag size */, HashType.SHA256);
    KeyData key1 = Registry.INSTANCE.newKeyData(template1);
    // No primary key.
    KeysetHandle keysetHandle = new KeysetHandle(Keyset.newBuilder()
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key1)
            .setKeyId(1)
            .setStatus(KeyStatusType.ENABLED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .build());
    // No primary key.
    try {
      Registry.INSTANCE.getPrimitives(keysetHandle);
      fail("Invalid keyset. Expect GeneralSecurityException");
    } catch (GeneralSecurityException e) {
      assertTrue(e.toString().contains("keyset doesn't contain a valid primary key"));
    }

    // Primary key is disabled.
    keysetHandle = new KeysetHandle(Keyset.newBuilder()
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key1)
            .setKeyId(1)
            .setStatus(KeyStatusType.DISABLED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .setPrimaryKeyId(1)
        .build());
    try {
      Registry.INSTANCE.getPrimitives(keysetHandle);
      fail("Invalid keyset. Expect GeneralSecurityException");
    } catch (GeneralSecurityException e) {
      assertTrue(e.toString().contains("keyset doesn't contain a valid primary key"));
    }
  }

  @Test
  public void testCustomKeyManagerHandling() throws Exception {
    // Create a keyset.
    KeyTemplate template1 =  TestUtil.createAesGcmKeyTemplate(16 /* key_size */);
    KeyTemplate template2 =  TestUtil.createAesCtrHmacAeadKeyTemplate(
        16 /* AES key size */,
        12 /* IV size */,
        16 /* HMAC key size */,
        10 /* tag size */);
    KeyData key1 = Registry.INSTANCE.newKeyData(template1);
    KeyData key2 = Registry.INSTANCE.newKeyData(template2);
    KeysetHandle keysetHandle = new KeysetHandle(Keyset.newBuilder()
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key1)
            .setKeyId(1)
            .setStatus(KeyStatusType.ENABLED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .addKey(Keyset.Key.newBuilder()
            .setKeyData(key2)
            .setKeyId(2)
            .setStatus(KeyStatusType.ENABLED)
            .setOutputPrefixType(OutputPrefixType.TINK)
            .build())
        .setPrimaryKeyId(2)
        .build());
    // Get a PrimitiveSet using registered key managers.
    PrimitiveSet<Aead> aeadSet = Registry.INSTANCE.getPrimitives(keysetHandle);
    List<PrimitiveSet<Aead>.Entry<Aead>> aead1List =
        aeadSet.getPrimitive(keysetHandle.getKeyset().getKey(0));
    List<PrimitiveSet<Aead>.Entry<Aead>> aead2List =
        aeadSet.getPrimitive(keysetHandle.getKeyset().getKey(1));
    assertEquals(1, aead1List.size());
    assertEquals(AesGcmJce.class, aead1List.get(0).getPrimitive().getClass());
    assertEquals(1, aead2List.size());
    assertEquals(EncryptThenAuthenticate.class, aead2List.get(0).getPrimitive().getClass());

    // Get a PrimitiveSet using a custom key manager for key1.
    KeyManager<Aead, MessageLite, MessageLite> customManager = new CustomAeadKeyManager();
    aeadSet = Registry.INSTANCE.getPrimitives(keysetHandle, customManager);
    aead1List = aeadSet.getPrimitive(keysetHandle.getKeyset().getKey(0));
    aead2List = aeadSet.getPrimitive(keysetHandle.getKeyset().getKey(1));
    assertEquals(1, aead1List.size());
    assertEquals(DummyAead.class, aead1List.get(0).getPrimitive().getClass());
    assertEquals(1, aead2List.size());
    assertEquals(EncryptThenAuthenticate.class, aead2List.get(0).getPrimitive().getClass());
  }

  // TODO(przydatek): Add more tests for creation of PrimitiveSets.
}

File: go/tink/registry_test.go
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package tink_test

import (
	"testing"

	"github.com/golang/protobuf/proto"
	"github.com/google/tink/go/aead"
	"github.com/google/tink/go/mac"
	subtleAEAD "github.com/google/tink/go/subtle/aead"
	subtleMac "github.com/google/tink/go/subtle/mac"
	"github.com/google/tink/go/testkeysethandle"
	"github.com/google/tink/go/testutil"
	"github.com/google/tink/go/tink"
	gcmpb "github.com/google/tink/proto/aes_gcm_go_proto"
	commonpb "github.com/google/tink/proto/common_go_proto"
	hmacpb "github.com/google/tink/proto/hmac_go_proto"
	tinkpb "github.com/google/tink/proto/tink_go_proto"
)

func TestRegisterKeyManager(t *testing.T) {
	// get HMACKeyManager
	_, err := tink.GetKeyManager(mac.HMACTypeURL)
	if err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	// get AESGCMKeyManager
	_, err = tink.GetKeyManager(aead.AESGCMTypeURL)
	if err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	// some random typeurl
	if _, err = tink.GetKeyManager("some url"); err == nil {
		t.Errorf("expect an error when a type url doesn't exist in the registry")
	}
}

func TestRegisterKeyManagerWithCollision(t *testing.T) {
	// dummyKeyManager's typeURL is equal to that of AESGCM
	var dummyKeyManager tink.KeyManager = new(testutil.DummyAEADKeyManager)
	// This should fail because overwriting is disallowed.
	err := tink.RegisterKeyManager(dummyKeyManager)
	if err == nil {
		t.Errorf("%s shouldn't be registered again", aead.AESGCMTypeURL)
	}

	km, err := tink.GetKeyManager(aead.AESGCMTypeURL)
	if err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	// This should fail because overwriting is disallowed, even with the same key manager.
	err = tink.RegisterKeyManager(km)
	if err == nil {
		t.Errorf("%s shouldn't be registered again", aead.AESGCMTypeURL)
	}
}

func TestNewKeyData(t *testing.T) {
	// new Keydata from a Hmac KeyTemplate
	keyData, err := tink.NewKeyData(mac.HMACSHA256Tag128KeyTemplate())
	if err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	if keyData.TypeUrl != mac.HMACTypeURL {
		t.Errorf("invalid key data")
	}
	key := new(hmacpb.HmacKey)
	if err := proto.Unmarshal(keyData.Value, key); err != nil {
		t.Errorf("unexpected error when unmarshal HmacKey: %s", err)
	}
	// nil
	if _, err := tink.NewKeyData(nil); err == nil {
		t.Errorf("expect an error when key template is nil")
	}
	// unregistered type url
	template := &tinkpb.KeyTemplate{TypeUrl: "some url", Value: []byte{0}}
	if _, err := tink.NewKeyData(template); err == nil {
		t.Errorf("expect an error when key template contains unregistered typeURL")
	}
}

func TestNewKey(t *testing.T) {
	// aead template
	aesGcmTemplate := aead.AES128GCMKeyTemplate()
	key, err := tink.NewKey(aesGcmTemplate)
	if err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	var aesGcmKey *gcmpb.AesGcmKey = key.(*gcmpb.AesGcmKey)
	aesGcmFormat := new(gcmpb.AesGcmKeyFormat)
	if err := proto.Unmarshal(aesGcmTemplate.Value, aesGcmFormat); err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	if aesGcmFormat.KeySize != uint32(len(aesGcmKey.KeyValue)) {
		t.Errorf("key doesn't match template")
	}
	//nil
	if _, err := tink.NewKey(nil); err == nil {
		t.Errorf("expect an error when key template is nil")
	}
	// unregistered type url
	template := &tinkpb.KeyTemplate{TypeUrl: "some url", Value: []byte{0}}
	if _, err := tink.NewKey(template); err == nil {
		t.Errorf("expect an error when key template is not registered")
	}
}

func TestPrimitiveFromKeyData(t *testing.T) {
	// hmac keydata
	keyData := testutil.NewHMACKeyData(commonpb.HashType_SHA256, 16)
	p, err := tink.PrimitiveFromKeyData(keyData)
	if err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	var _ *subtleMac.HMAC = p.(*subtleMac.HMAC)
	// unregistered url
	keyData.TypeUrl = "some url"
	if _, err := tink.PrimitiveFromKeyData(keyData); err == nil {
		t.Errorf("expect an error when typeURL has not been registered")
	}
	// unmatched url
	keyData.TypeUrl = aead.AESGCMTypeURL
	if _, err := tink.PrimitiveFromKeyData(keyData); err == nil {
		t.Errorf("expect an error when typeURL doesn't match key")
	}
	// nil
	if _, err := tink.PrimitiveFromKeyData(nil); err == nil {
		t.Errorf("expect an error when key data is nil")
	}
}

func TestPrimitive(t *testing.T) {
	// hmac key
	key := testutil.NewHMACKey(commonpb.HashType_SHA256, 16)
	serializedKey, _ := proto.Marshal(key)
	p, err := tink.Primitive(mac.HMACTypeURL, serializedKey)
	if err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	var _ *subtleMac.HMAC = p.(*subtleMac.HMAC)
	// unregistered url
	if _, err := tink.Primitive("some url", serializedKey); err == nil {
		t.Errorf("expect an error when typeURL has not been registered")
	}
	// unmatched url
	if _, err := tink.Primitive(aead.AESGCMTypeURL, serializedKey); err == nil {
		t.Errorf("expect an error when typeURL doesn't match key")
	}
	// void key
	if _, err := tink.Primitive(aead.AESGCMTypeURL, nil); err == nil {
		t.Errorf("expect an error when key is nil")
	}
	if _, err := tink.Primitive(aead.AESGCMTypeURL, []byte{}); err == nil {
		t.Errorf("expect an error when key is nil")
	}
	if _, err := tink.Primitive(aead.AESGCMTypeURL, []byte{0}); err == nil {
		t.Errorf("expect an error when key is nil")
	}
}

func TestPrimitives(t *testing.T) {
	// valid input
	template1 := aead.AES128GCMKeyTemplate()
	template2 := aead.AES256GCMKeyTemplate()
	keyData1, _ := tink.NewKeyData(template1)
	keyData2, _ := tink.NewKeyData(template2)
	keyset := tink.CreateKeyset(2, []*tinkpb.Keyset_Key{
		tink.CreateKey(keyData1, tinkpb.KeyStatusType_ENABLED, 1, tinkpb.OutputPrefixType_TINK),
		tink.CreateKey(keyData2, tinkpb.KeyStatusType_ENABLED, 2, tinkpb.OutputPrefixType_TINK),
	})
	handle, _ := testkeysethandle.KeysetHandle(keyset)
	ps, err := tink.Primitives(handle)
	if err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	var aesGcm *subtleAEAD.AESGCM = ps.Primary.Primitive.(*subtleAEAD.AESGCM)
	if len(aesGcm.Key) != 32 {
		t.Errorf("primitive doesn't match input keyset handle")
	}
	// custom manager
	customManager := new(testutil.DummyAEADKeyManager)
	ps, err = tink.PrimitivesWithKeyManager(handle, customManager)
	if err != nil {
		t.Errorf("unexpected error: %s", err)
	}
	var _ *testutil.DummyAEAD = ps.Primary.Primitive.(*testutil.DummyAEAD)
	// keysethandle is nil
	if _, err := tink.Primitives(nil); err == nil {
		t.Errorf("expect an error when keysethandle is nil")
	}
	// keyset is empty
	keyset = tink.CreateKeyset(1, []*tinkpb.Keyset_Key{})
	handle, _ = testkeysethandle.KeysetHandle(keyset)
	if _, err := tink.Primitives(handle); err == nil {
		t.Errorf("expect an error when keyset is empty")
	}
	keyset = tink.CreateKeyset(1, nil)
	handle, _ = testkeysethandle.KeysetHandle(keyset)
	if _, err := tink.Primitives(handle); err == nil {
		t.Errorf("expect an error when keyset is empty")
	}
	// no primary key
	keyset = tink.CreateKeyset(3, []*tinkpb.Keyset_Key{
		tink.CreateKey(keyData1, tinkpb.KeyStatusType_ENABLED, 1, tinkpb.OutputPrefixType_TINK),
		tink.CreateKey(keyData2, tinkpb.KeyStatusType_ENABLED, 2, tinkpb.OutputPrefixType_TINK),
	})
	handle, _ = testkeysethandle.KeysetHandle(keyset)
	if _, err := tink.Primitives(handle); err == nil {
		t.Errorf("expect an error when there is no primary key")
	}
	// there is primary key but it is disabled
	keyset = tink.CreateKeyset(1, []*tinkpb.Keyset_Key{
		tink.CreateKey(keyData1, tinkpb.KeyStatusType_DISABLED, 1, tinkpb.OutputPrefixType_TINK),
		tink.CreateKey(keyData2, tinkpb.KeyStatusType_ENABLED, 2, tinkpb.OutputPrefixType_TINK),
	})
	handle, _ = testkeysethandle.KeysetHandle(keyset)
	if _, err := tink.Primitives(handle); err == nil {
		t.Errorf("expect an error when primary key is disabled")
	}
	// multiple primary keys
	keyset = tink.CreateKeyset(1, []*tinkpb.Keyset_Key{
		tink.CreateKey(keyData1, tinkpb.KeyStatusType_ENABLED, 1, tinkpb.OutputPrefixType_TINK),
		tink.CreateKey(keyData2, tinkpb.KeyStatusType_ENABLED, 1, tinkpb.OutputPrefixType_TINK),
	})
	handle, _ = testkeysethandle.KeysetHandle(keyset)
	if _, err := tink.Primitives(handle); err == nil {
		t.Errorf("expect an error when there are multiple primary keys")
	}
}
